src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+}
+
+GroupDataProvider::GroupMappingIterator * GroupDataProviderImpl::IterateGroupMappings(chip::FabricIndex fabric_index,
+                                                                                      EndpointId endpoint_id)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mEndpointIterators.CreateObject(*this, fabric_index, endpoint_id);
+}
+
+GroupDataProviderImpl::GroupMappingIteratorImpl::GroupMappingIteratorImpl(GroupDataProviderImpl & provider,
+                                                                          chip::FabricIndex fabric_index,
+                                                                          chip::EndpointId endpoint_id) :
+    mProvider(provider),
+    mFabric(fabric_index), mEndpoint(endpoint_id)
+{
+    FabricData fabric(fabric_index);
+    ReturnOnFailure(fabric.Load(provider.mStorage));
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    uint16_t count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        ReturnOnFailure(endpoint.Load(provider.mStorage));
+        if (endpoint.id == endpoint_id)
+        {
+            // Target endpoint found
+            mGroup = endpoint.first_group;
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while (++count < fabric.endpoint_count);
+}
+
+size_t GroupDataProviderImpl::GroupMappingIteratorImpl::Count()
+{
+    size_t count = 0;
+
+    GroupData group(mFabric, mEndpoint, mGroup);
+    chip::GroupId prev_gid = kUndefinedGroupId;
+
+    while ((kUndefinedGroupId != group.group) && (prev_gid != group.group))
+    {
+        if (CHIP_NO_ERROR != group.Load(mProvider.mStorage))
+        {
+            break;
+        }
+        prev_gid    = group.group;
+        group.group = group.next;
+        count++;
+    }
+    return count;
+}
+
+bool GroupDataProviderImpl::GroupMappingIteratorImpl::Next(GroupMapping & item)
+{
+    GroupData group(mFabric, mEndpoint, mGroup);
+    CHIP_ERROR err = group.Load(mProvider.mStorage);
+    if ((kUndefinedGroupId == mGroup) || (CHIP_NO_ERROR != err))
+    {
+        return false;
+    }
+    item.endpoint = mEndpoint;
+    item.group    = mGroup;
+    strncpy(item.name, group.name, GroupData::kGroupNameMax);
+    mGroup = group.next;
+    return true;
+}
+
+void GroupDataProviderImpl::GroupMappingIteratorImpl::Release()
+{
+    mProvider.mEndpointIterators.ReleaseObject(this);
+}
+
+//
+// Group States
+//
+
+CHIP_ERROR GroupDataProviderImpl::SetGroupState(size_t state_index, const GroupState & in_state)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(IsFabricGroupId(in_state.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    StateListData states;
+
+    if (CHIP_NO_ERROR != states.Load(mStorage))
+    {
+        // First state
+        VerifyOrReturnError(0 == state_index, CHIP_ERROR_INVALID_ARGUMENT);
+        states.first = 1; // Arbitrary index > 0
```

Comment:
I am struggling a bit with the multiple indices used, some of which seem to hold 0 as a special value. I am having a hard time determining if there are off-by-ones in conditions such as `(++actual_index < states.count)` and when there are comparisons between "real" and "pseudo" indices such as: `if (new_index == state.index)` and `(actual_index == state_index)`

Reply:
Making the linked-list index > 0 makes easier to check for previous entries, but these "indexes" could have any value, so long they're unique. To avoid the confusion, I have rename the linked-list "index" as "id", and all mentions of the actual list index is simply called "index". I have also removed the > 0 requirement.